### PR TITLE
Fix gamma mixture

### PR DESCRIPTION
**Pull request by albertpod**
Initially, this PR aimed to solve #84.

It enables importance sampling for the `shape` parameters of the Gamma Mixture node (see supporting notebook on `Gamma Mixture`).

We have filled in the missing functions for `ImportanceSamplingApproximation`.

We have added the missing `rand` function for the `GammaShapeRate` type.

Average energy functions for `GammaShapeRate` were fused into a single function.

### CODE REVIEW

File: src/approximations/importance.jl [hunk 1]
```diff
@@ -13,7 +13,7 @@ convenient functions to generate samples and weights to approximate expectations
 - `rng`: random number generator objects
 - `nsamples`: number of samples generated by default
 """
-struct ImportanceSamplingApproximation{T, R}
+struct ImportanceSamplingApproximation{T, R} <: AbstractFormConstraint
```
Comment (bvdmitri) [inline src/approximations/importance.jl:16]: The `ImportanceSamplingApproximation` should not be declared as an `AbstractFormConstraint`. This is wrong from the semantic point of view. There should be an extra `MomentsApproximationFormConstraint` (or something similar).

File: src/approximations/importance.jl [hunk 2]
```diff
@@ -83,3 +87,20 @@ function approximate_meancov(approximation::ImportanceSamplingApproximation, g::
...
+is_point_mass_form_constraint(::ImportanceSamplingApproximation) = false
+default_form_check_strategy(::ImportanceSamplingApproximation)   = FormConstraintCheckEach()
+default_prod_constraint(::ImportanceSamplingApproximation)       = ProdGeneric()
+
+function constrain_form(approximation::ImportanceSamplingApproximation, distribution::DistProduct)
+    m, v = approximate_meancov(approximation, x -> pdf(getright(distribution), x), getleft(distribution))
+
+    a = m^2 / v
+    b = m / v
+    return GammaShapeRate(a, b)
```
Comment (bvdmitri) [inline src/approximations/importance.jl:100]: This should be changed somehow

File: src/approximations/importance.jl [hunk 3]
```diff
@@ -83,3 +87,20 @@ function approximate_meancov(approximation::ImportanceSamplingApproximation, g::
...
+end
+
+constrain_form(::ImportanceSamplingApproximation, distribution::Any) = distribution
+
+make_form_constraint(::Type{ImportanceSamplingApproximation}, args...; kwargs...) =
```
Comment (bvdmitri) [inline src/approximations/importance.jl:105]: This should be removed in favor of separate `MomentsApproximationFormConstraint `

File: src/distributions/gamma_shape_rate.jl [hunk 4]
```diff
@@ -65,3 +65,7 @@ end
 
 Distributions.pdf(dist::GammaShapeRate, x::Real)    = (rate(dist)^shape(dist)) / gamma(shape(dist)) * x^(shape(dist) - 1) * exp(-rate(dist) * x)
 Distributions.logpdf(dist::GammaShapeRate, x::Real) = shape(dist) * log(rate(dist)) - loggamma(shape(dist)) + (shape(dist) - 1) * log(x) - rate(dist) * x
+
+function Random.rand(rng::AbstractRNG, dist::GammaShapeRate)
```
Comment (bvdmitri) [inline src/distributions/gamma_shape_rate.jl:69]: Lets also add more `rand` methods (which accepts size and inplace versions).

File: src/nodes/gamma_shape_rate.jl [hunk 5]
```diff
@@ -4,10 +4,8 @@ import SpecialFunctions: loggamma
...
 @node GammaShapeRate Stochastic [out, (α, aliases = [shape]), (β, aliases = [rate])]
 
-@average_energy GammaShapeRate (q_out::Any, q_α::PointMass, q_β::Any) = begin
-    mean(loggamma, q_α) - mean(q_α) * mean(log, q_β) - (mean(q_α) - 1.0) * mean(log, q_out) + mean(q_β) * mean(q_out)
-end
-
-@average_energy GammaShapeRate (q_out::Any, q_α::GammaDistributionsFamily, q_β::Any) = begin
-    mean(loggamma, q_α) - mean(q_α) * mean(log, q_β) - (mean(q_α) - 1.0) * mean(log, q_out) + mean(q_β) * mean(q_out)
-end
+@average_energy GammaShapeRate (q_out::Any, q_α::Union{PointMass, SampleList, GammaDistributionsFamily}, q_β::Any) =
```
Comment (bvdmitri) [inline src/nodes/gamma_shape_rate.jl:7]: I wonder if can simply put `Any` there

**Review by bvdmitri — CHANGES_REQUESTED** [review]

File: src/approximations/importance.jl (hunk 1 above)
Comment (albertpod) [reply src/approximations/importance.jl:16]: @ismailsenoz @bartvanerp

**Pull request closed without merging**